File: web/src/lib/relay.ts
/**
 * Blind relay (CHA-22) — the browser twin of the cloud store's PRIVATE tier.
 *
 * This models the cloud in the multi-device demo. It holds nothing but
 * ciphertext records: {id, ciphertext, iv, created_at}. There is no field for a
 * key and no code path that accepts one — push() copies out only those four
 * fields and hard-rejects anything that smells like key material. That makes
 * "the cloud never held the key" a structural property you can read off the code,
 * not a promise. Device B pulls a record and decrypts locally with the key it got
 * over the QR channel — the relay is never in the loop for that.
 */

import type { EncryptedPayload } from './crypto';

export interface RelayRecord {
	id: string;
	ciphertext: string; // base64url
	iv: string; // base64url 96-bit nonce
	created_at: string; // ISO 8601
}

/** Property names we refuse to store — a tripwire against leaking key material. */
const FORBIDDEN_KEYS = ['key', 'k', 'secret', 'privatekey', 'private_key', 'aeskey'];

export class BlindRelay {
	#records = new Map<string, RelayRecord>();

	/**
	 * Store one ciphertext record. Only {id, ciphertext, iv, created_at} are kept;
	 * any extra property (especially anything key-shaped) is rejected outright.
	 */
	push(rec: RelayRecord): void {
		for (const prop of Object.keys(rec)) {
			if (FORBIDDEN_KEYS.includes(prop.toLowerCase())) {
				throw new Error(`[relay] refused: record carries forbidden field "${prop}"`);
			}
		}
		if (!rec.id || !rec.ciphertext || !rec.iv) {
			throw new Error('[relay] record must have id, ciphertext and iv');
		}
		// Copy out ONLY the blind fields — never trust the caller's object wholesale.
		this.#records.set(rec.id, {
			id: rec.id,
			ciphertext: rec.ciphertext,
			iv: rec.iv,
			created_at: rec.created_at
		});
	}

	/** Convenience: push from an {id} + EncryptedPayload produced by crypto.ts. */
	pushEncrypted(id: string, payload: EncryptedPayload, created_at: string): void {
		this.push({ id, ciphertext: payload.ciphertext, iv: payload.iv, created_at });
	}

	/** Fetch a ciphertext record — the only thing the relay can ever return. */
	pull(id: string): RelayRecord | null {
		const rec = this.#records.get(id);
		return rec ? { ...rec } : null;
	}

	list(): RelayRecord[] {
		return [...this.#records.values()].map((r) => ({ ...r }));
	}

	clear(): void {
		this.#records.clear();
	}

	/**
	 * Structural self-report for the demo's relay inspector. `holdsKey` is always
	 * false by construction — there is no key field to hold.
	 */
	inspect(): { count: number; fields: string[]; holdsKey: false } {
		return {
			count: this.#records.size,
			fields: ['id', 'ciphertext', 'iv', 'created_at'],
			holdsKey: false
		};
	}
}

/** Shared relay instance for the multi-device demo (both device panels use it). */
export const relay = new BlindRelay();

File: server/relay.py
"""Blind relay + QR key-envelope for the multi-device flow (CHA-22).

The Python twin of ``web/src/lib/relay.ts`` + ``keyenvelope.ts``. It exists so the
multi-device claim is provable headlessly (see ``server/verify_cha22.py`` and
``tests/test_multidevice.py``), in the same runtime as the MCP server that already
decrypts PRIVATE cards.

Two moving parts, deliberately separate:

  * ``BlindRelay`` — models the cloud. Holds only ciphertext records
    ``{id, ciphertext, nonce, created_at}``. There is no key field and no code
    path that accepts one; ``push`` copies out just those four fields and rejects
    anything key-shaped. "The cloud never held the key" is thus structural.

  * key envelope — the versioned wrapper the QR carries out-of-band between the
    user's own devices. It never touches the relay.
"""
from __future__ import annotations

import base64
import json
from dataclasses import dataclass

ENVELOPE_VERSION = 1
_ALG = "AES-256-GCM"
_KEY_BYTES = 32  # 256-bit

# Property names the relay refuses to store — a tripwire against key leakage.
_FORBIDDEN_FIELDS = {"key", "k", "secret", "privatekey", "private_key", "aeskey"}


# ── key envelope ──────────────────────────────────────────────────────────────

def _b64url_len(b64: str) -> int:
    rem = len(b64) % 4
    padded = b64 + "=" * (4 - rem) if rem else b64
    return len(base64.urlsafe_b64decode(padded))


def _is_valid_key_b64(b64: str) -> bool:
    try:
        return _b64url_len(b64) == _KEY_BYTES
    except Exception:
        return False


def build_key_envelope(key_b64: str) -> str:
    """Wrap a base64url key into the versioned envelope string a QR will carry."""
    if not _is_valid_key_b64(key_b64):
        raise ValueError("refusing to wrap a non-32-byte key")
    return json.dumps({"v": ENVELOPE_VERSION, "alg": _ALG, "k": key_b64})


def parse_key_envelope(payload: str) -> str:
    """Parse a scanned/pasted QR payload back to a base64url key.

    Accepts the JSON envelope or a bare base64url key. Raises on anything that
    does not yield a valid 32-byte key.
    """
    text = payload.strip()
    if not text:
        raise ValueError("empty payload")

    if text.startswith("{"):
        try:
            env = json.loads(text)
        except json.JSONDecodeError as exc:
            raise ValueError("malformed envelope JSON") from exc
        alg = env.get("alg")
        if alg and alg != _ALG:
            raise ValueError(f"unsupported alg: {alg}")
        k = env.get("k")
        if not isinstance(k, str) or not _is_valid_key_b64(k):
            raise ValueError("envelope has no valid key")
        return k

    if not _is_valid_key_b64(text):
        raise ValueError("not a valid 32-byte base64url key")
    return text


# ── blind relay ────────────────────────────────────────────────────────────────

@dataclass
class RelayRecord:
    id: str
    ciphertext: str
    nonce: str
    created_at: str


class BlindRelay:
    """In-memory ciphertext-only relay. Never accepts or returns a key."""

    def __init__(self) -> None:
        self._records: dict[str, RelayRecord] = {}

    def push(self, *, id: str, ciphertext: str, nonce: str, created_at: str, **extra) -> None:
        """Store one ciphertext record. Any key-shaped extra field is rejected."""
        for prop in extra:
            if prop.lower() in _FORBIDDEN_FIELDS:
                raise ValueError(f'refused: record carries forbidden field "{prop}"')
        if extra:
            raise ValueError(f"unexpected fields on relay record: {sorted(extra)}")
        if not id or not ciphertext or not nonce:
            raise ValueError("record must have id, ciphertext and nonce")
        self._records[id] = RelayRecord(
            id=id, ciphertext=ciphertext, nonce=nonce, created_at=created_at
        )

    def pull(self, id: str) -> RelayRecord | None:
        rec = self._records.get(id)
        return RelayRecord(**vars(rec)) if rec else None

    def list(self) -> list[RelayRecord]:
        return [RelayRecord(**vars(r)) for r in self._records.values()]

    def clear(self) -> None:
        self._records.clear()

    def inspect(self) -> dict:
        """Structural self-report — ``holds_key`` is False by construction."""
        return {
            "count": len(self._records),
            "fields": ["id", "ciphertext", "nonce", "created_at"],
            "holds_key": False,
        }
